Feature: 게시판 페이지 퍼블리싱

Monixc opened the pull request:
### 📌 PR 종류 (PR Type)
<!-- 아래 항목 중 해당하는 곳에 [x] 로 표시해주세요. -->

- [ ] 🐛 버그 수정 (Bugfix)
- [x] ✨ 기능 추가 (Feature)
- [x] 🎨 UI 스타일 변경 (Style) - CSS, 클래스명, 레이아웃 등 UI 변경
- [x] ♻️ 코드 리팩토링 (Refactor) - 기능 변화 없이 코드 구조 개선
- [x] 📦 패키지/빌드 관련 변경 (Build) - 설정 파일 수정, 의존성 추가 등
- [ ] 🧱 테스트 코드 추가/수정 (Test)
- [ ] 📝 문서 수정 (Docs)
- [ ] 🚧 WIP (작업 중 PR)
- [ ] 기타 (Other): <!-- 직접 입력 -->

---

### 📝 요약 (Summary)
<!-- 한두 문장으로 변경 내용을 요약해주세요. -->
React Router를 사용해 URL 기반 게시판 퍼블리싱 구현했습니다.
---

### 🔍 상세 내용 (Describe your changes)
<!-- 어떤 변경사항이 있었는지, 왜 변경했는지 상세히 작성해주세요. 코드 흐름, 주요 로직 등을 설명해도 좋습니다. -->

**url  기반 게시판 ui 구현**
- shadcn/ui의 tabs, card, badge 컴포넌트 활용하여 게시판 페이지의 전체적인 레이아웃을 퍼블리싱했습니다.
- 사용자가 탭을 클릭하면 url이 변경되고, 해당 url에 맞는 카테고리가 활성화되는  ui를 구현했습니다.

**컴포넌트**
- `BoardTabs.tsx`: 카테로리 탭 네이게이션 UI
- `ListItem.tsx`: 개별 게시물을 나타내는 카드 컴포넌트

**React Router 설정**
- `createBrwserRouter`를 사용하여 애플리케이션 라우팅 구조 설정했습니다.
- 각 게시판 카테고리가 고유 url을 가져, 주소로 접근가능합니다.

**가상 api 연동**
- 초기 ui 및 데이터 흐름 확인을 위해 categoryId에 따라 해당하느 게시물 목록을 반환하는 가상 API를 만들어 사용합니다.
---



### 📸 스크린샷 (필수)
<!-- UI 변경이 있는 경우, 변경 전/후 비교나 주요 화면을 첨부해주세요. -->
<img width="1788" height="1047" alt="image" src="https://github.com/user-attachments/assets/25790be1-21a0-46a6-a3f9-d01d7d0d58d8" />

---

### ✅ 확인 사항 (CheckList)
<!-- PR을 보내기 전 스스로 점검할 수 있도록 합니다. 팀원이 리뷰할 때도 참고할 수 있습니다. -->
- [x] 브라우저에서 정상 동작 확인
- [x] 기존 기능에 영향 없음
- [x] 반응형 UI 대응 확인
- [x] 코드 컨벤션에 맞게 작성됨 (eslint, prettier 등)
- [ ] 테스트 코드 추가/수정 또는 테스트 통과 확인

---
### 💬 기타 참고사항 (Optional)
<!-- 리뷰어가 알면 좋을 추가 정보, 요청 사항 등을 작성해주세요. -->
ex) 이 부분 로직이 좀 복잡해서 한번 더 확인 부탁드립니다 🙏

- 가상 api를 활용하여 게시판 이동을 구현하였으나 api 연동 과정에서 실제 api 구조(post/categoryId, recruits/categoryId)에 따라 변경 예정입니다. 흐름 파악 용도로만 확인해주세요. 
- BoardTab 구현 과정에서 radixui 우선순위 때문인지 스타일링이 적용되지 않아 `!`를 사용했습니다. `!` 사용이 적절한지, 더 나은 방법이 있는지 의견 부탁드립니다.
- main, app, 컴포넌트 구조가 적절한지 확인 부탁드립니다.
- BoardTab에서 직접 필터링하는 방식을 사용하다가, api 연동할 것을 생각해 가상 api를 사용하고 라우팅하는 방식으로 중간에 변경했습니다. [1a139ab](https://github.com/prgrms-fullstack-devcourse/Comeet_FE/pull/3/commits/1a139ab6cae5a63567e35b73fd1437798368a74d)부터 확인해주세요!

Monixc committed:
feat: tabs, card, badge 컴포넌트 추가가

Monixc committed:
feat: BoardTabs 컴포넌트 추가

- 게시물 카테고리에 따라 탭을 생성하고 필터링
- 전체, 자유, 질문, 프로젝트, 모각코 탭 추가

Monixc committed:
feat: ListItem 컴포넌트 추가

- 게시물 정보를 표시하는 ListItem 컴포넌트 구현

Monixc committed:
feat: Post 인터페이스 추가

- 게시물 정보를 정의하는 Post 인터페이스 구현

Monixc committed:
feat: BoardPage 컴포넌트 추가

- 게시물 목록을 표시하는 BoardPage 컴포넌트 구현

Monixc committed:
style: GlobalLayout 컴포넌트 코드 정리

- floating 버튼 정렬을 위해 relative 설정

Monixc committed:
Merge branch 'develop' of https://github.com/prgrms-fullstack-devcourse/Comeet_FE into feature/page-main

Monixc committed:
fix: auto-label.yml에서 GitHub API 호출 수정

- `github.issues.addLabels`를 `github.rest.issues.addLabels`로 변경하여 최신 API 사용

Monixc committed:
style: ListItem 컴포넌트에 hover 효과 추가

- 카드에 hover 시 배경색 변화 효과를 추가하여 사용자 경험 개선

Monixc committed:
style: merge conflict 과정에서 누락된 relative 속성 복구

Monixc committed:
chore: BoardPage 컴포넌트를 home 디렉토리로 이동

- BoardPage 컴포넌트를 src/pages/home/index.tsx로 이동하여 구조 개선

Monixc committed:
chore: react-router-dom 및 관련 타입 추가

Monixc committed:
feat: react-router-dom을 사용하여 라우팅 설정 추가

- App 컴포넌트에서 BoardPage를 Outlet으로 변경하여 하위 라우팅을 지원
- main.tsx에서 createBrowserRouter를 사용하여 기본 경로와 로그인 경로 설정

Monixc committed:
feat: 가상의 API 함수 추가 및 게시물 데이터 정의

- 전체 게시물 목록을 포함하는 ALL_POSTS 배열 추가
- categoryId에 따라 게시물 목록을 가져오는 fetchPosts 함수 구현

Monixc committed:
feat: 게시물 카테고리 관리 및 UI 개선

- tab에서 필터링 하던 것을 가상 api에서 가져오는 categoryid에 따라 페이지 라우팅하여 표시하도록 변경

Monixc committed:
fix: auto-label.yml 파일에 권한 설정 추가

- issues 및 pull-requests에 대한 쓰기 권한을 추가하여 자동 라벨링 오류 수정

Monixc committed:
fix: auto-label.yml 파일의 라벨이 이모지 없이 생성되는 문제 수정

- 각 라벨에 이모지를 추가하여 가독성을 향상시킴

Monixc commented:
auto-label.yml 오류 해결하여 이제 문제 없이 동작합니다. 작업 중 입력한 커밋 타입에 따라 자동으로 설정됩니다.

woo-dev-log commented:
가상 API 활용한 점진적 개발 방식이 체계적이고, 라우팅 구조도 API 연동을 잘 고려해서 구현하신 것 같아요.
전반적으로 깔끔하게 구현하신 것 같습니다! 👍

woo-dev-log reviewed (approved):
가상 API 활용한 점진적 개발 방식이 체계적이고, 라우팅 구조도 API 연동을 잘 고려해서 구현하신 것 같아요.
전반적으로 깔끔하게 구현하신 것 같습니다! 👍

ksy90101 commented on src/components/common/BoardTabs.tsx:
```diff
@@ -0,0 +1,44 @@
+
+interface BoardTabsProps {
+  posts: Post[];
+  currentCategory: string;
```
정해진 Category가 있을것이라고 생각이 듭니다.
"BOARD_CATEGORIES" 이거겠네요.
따라서 단순히 타입이 string이 아닐것으로 기대됩니다.

타입을 최대한 좁히는걸 추천드립니다.

Monixc replied:
```
export const BOARD_CATEGORIES = [
  { value: "all", label: "전체" },
  { value: "1", label: "자유" },
  { value: "2", label: "질문" },
  { value: "3", label: "프로젝트" },
  { value: "4", label: "모각코" },
] as const;

export type UICategory = typeof BOARD_CATEGORIES[number]["value"]; 
```
as const 사용하여 리터럴 타입으로 선언하고, UICategory 타입으로 최대한 타입 좁히는 방식으로 구현해보겠습니다.

ksy90101 commented on src/components/common/ListItem.tsx:
```diff
@@ -0,0 +1,41 @@
+
+export const ListItem = ({ post }: ListItemProps) => {
+  const categoryLabel =
+    BOARD_CATEGORIES.find((c) => c.value === post.categoryId.toString())
+      ?.label || "";
```
category가 따로 DB 테이블에 있는것으로 예측되는데.
결국 category가 추가되면 프론트의 BOARD_CATEGORIES 해당 코드도 변경되어야 하는 문제가 있어 보여요.
테이블로 따로 뺀다는건 동적으로 관리한다는것이고, 상수를 한다는것은 정적으로 관리한다는것인데
지금 둘다 쓰고 있는데 둘다의 장단점을 하나도 취하고 쓰지 않고 있습니다.
따라서 이 부분의 설계를 다시 진행해주세요.

Monixc replied:
현재 ERD에서는 category_id만 사용하고 있어, 프론트에서는 메뉴 이름 출력을 위해 상수를 활용한 매핑 방식으로 처리하고 있었습니다. 설계 당시 게시판 확장을 고려하지 않고 고정된 4개 카테고리만을 상정해 이 방식도 충분하도고 생각했습니다.

하지만 멘토님 코멘트를 통해, 동적인 데이터 관리나 기능 확장을 고려했을 때 현재 구조가 비효율적이고, 두 방식의 장점을 모두 취하지 문제가 있다는걸 인지하게 되었습니다. 

이 부분은 백엔드측에 category 이름 정보를 함께 관리하도록 구조 개선 요청하겠습니다. 감사합니다

ksy90101 commented on src/lib/api.ts:
이번 프로젝트에서는 msw[https://tech.kakao.com/posts/458] 를 써보시면 어떨까요?

Monixc replied:
넵 msw 적용해보겠습니다.

ksy90101 commented on src/pages/home/index.tsx:
```diff
@@ -0,0 +1,39 @@
+  const { category = "all" } = useParams<{ category: string }>();
+  const [posts, setPosts] = useState<Post[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<Error | null>(null);
```
해당 에러는 쓰이는곳이 없어 보이는데
error까지 state에 저장할 이유가 무엇인가요?

Monixc replied:
error, loading state는 게시판 출력을 api로 받아오는 것처럼 보이기 위해 작성한 코드의 흔적입니다.

commit 전 코드에서는 setTimeout을 사용해 비동기 흐름을 시뮬레이션 했고, 두 상태를 사용했었습니다. 과도한 처리인 듯 하여 해당 로직을 제거하였으나, 그 과정에서 상태 관리 코드 일부가 남았습니다.

당장 사용하는 상태는 아니지만 추후 api를 연동하는 과정에서는 loading, error 상태 관리가 필요할 듯 하여 유지하려고 합니다.

ksy90101 replied:
이거 react query 쓰면 다 해결됩니다.
react query 도입이번에 하시죠!

Monixc replied:
넵 react query 도입해보겠습니다!

ksy90101 commented on src/pages/home/index.tsx:
```diff
@@ -0,0 +1,39 @@
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<Error | null>(null);
+
+  useEffect(() => {
```
이번 api통신은 react query를 사용하시는걸 추천드립니다.

Monixc commented on src/components/common/BoardTabs.tsx:
```diff
@@ -0,0 +1,44 @@
+          <TabsTrigger
+            key={tab.value}
+            value={tab.value}
+            className="rounded-full border-brand-surface data-[state=active]:!bg-transparent data-[state=active]:!text-brand-primary data-[state=active]:!border-2 data-[state=active]:!border-brand-primary">
```
shadcn ui 기본 스타일 우선순위 때문에 brand- 컬러가 적용되지 않는 문제가 있었습니다. brand- 컬러를 우선으로 렌더링 하도록`! imoprttant` 사용하였습니다.

오늘 팀 회의 때 본 문제와 관련하여 회의하였고, Tab처럼 재사용이 많지 않은 컴포넌트는 `! important` 사용을 지양하고, shadcn ui 컴포넌트를 직접 수정하여 사용하기로 결정했습니다.

Monixc commented on src/components/common/ListItem.tsx:
```diff
@@ -0,0 +1,41 @@
+        </span>
+        <div className="flex items-center space-x-4">
+          <div className="flex items-center space-x-1">
+            <Heart className="size-4" />
```
웹 접근성 고려해서 수정하겠습니다.

jeongsarang-0701 commented:
필요한 기능들이 적절히 구현된 것 같습니다! 👍

Monixc committed:
style: UI 컴포넌트 개선 및 접근성 향상

- BoardTabs 컴포넌트의 클래스 수정으로 스타일 개선
- ListItem 컴포넌트에 aria-label 추가로 접근성 향상
- Tabs 컴포넌트의 코드 포맷팅 정리
- BOARD_CATEGORIES 상수 정의 방식 변경 및 UICategory 타입 추가
- fetchPosts 함수에서 디버깅 로그 제거

Monixc commented:
리뷰 반영하여 코드 수정했습니다: 
- BoardTabs 컴포넌트의 클래스 수정으로 스타일 개선
- ListItem 컴포넌트에 aria-label 추가로 접근성 향상
- Tabs 컴포넌트의 코드 포맷팅 정리
- BOARD_CATEGORIES 상수 정의 방식 변경 및 UICategory 타입 추가
- fetchPosts 함수에서 디버깅 로그 제거
